Add machine-readable property definitions to repository

rartino opened the pull request:
This PR adds machine-readable property definitions for all the OPTIMADE standardized properties using the format we developed over the last couple of workshops.

Old PR text:

~~This PR has started the work of adding property definitions to the repository on the format that is now in the specification.~~

~~I've opened it as a pull request even though it is not yet finished to:~~
~~- Make it more visible for discussions about the format.~~
~~- To enable people to look things over while definitions are being added.~~
~~- Possibly get help from others ("maintainers" should be able to push to the PR branch of my repo)~~ 

~~Note: the referenced command: "make properties" is not yet in place.~~

ml-evs commented on properties/structures/elements.yaml:
```diff
@@ -0,0 +1,31 @@
+    - MUST refer to the same elements in the same order, and therefore be of the same length, as elements_ratios, if the latter is provided.
+    - Note: This property SHOULD NOT contain the string "X" to indicate non-chemical elements or "vacancy" to indicate vacancies (in contrast to the field chemical_symbols for the species property).
+
+    Queryinig:
```
```suggestion
    Querying:
```

vaitkus commented on properties/REAMDE.md:
```diff
@@ -0,0 +1,13 @@
+The subdirectories below contain source files for property definitions for the standard OPTIMADE properties.
+They are organized with one subdirectory per OPTIMADE endpoint, and one directory `common` for common definitions used across multiple endpoints.
+
+The source files are formatted in yaml to be human-editable and uses JSON Schema pointers to to avoid duplication of information.
```
```suggestion
The source files are formatted in yaml to be human-editable and uses JSON Schema pointers to avoid duplication of information.
```

vaitkus replied:
```suggestion
The source files are formatted in YAML to be human-editable and use JSON Schema pointers to to avoid duplication of information.
```

vaitkus commented on properties/common/element.yaml:
```diff
@@ -0,0 +1,12 @@
+    The chemical symbol of an element
+
+    Requirements/Conventions:
+      - The strings are the chemical symbols, i.e., either a single uppercase letter or an uppercase letter followed by a number of lowercase letters.
```
```suggestion
      - The strings are the chemical symbols, i.e., either a single uppercase letter or an uppercase letter followed by no more than two lowercase letters.
```

rartino replied:
Should we edit the textual definition strings in this PR? I have a feeling this opens up this rather technical conversion to a lot of discussion about definition texts? Maybe better to do this as a separate PR?

vaitkus replied:
Ok, I can retract the last three comments if they seems out of place.

rartino replied:
Ok, good. Keep a note about these somewhere and we can edit them when (if?) we remove the parallel definitions in optimade.rst. But, right, we probably do not want the definitions to diverge between the two sources during conversion into YAML...

vaitkus commented on properties/structures/elements.yaml:
```diff
@@ -0,0 +1,31 @@
+    The chemical symbols of the different elements present in the structure.
+
+    Requirements/Conventions:
+    - The strings are the chemical symbols, i.e., either a single uppercase letter or an uppercase letter followed by a number of lowercase letters.
```
```suggestion
    - The strings are the chemical symbols, i.e., either a single uppercase letter or an uppercase letter followed by no more than two lowercase letters.
```

vaitkus commented on properties/structures/nelements.yaml:
```diff
@@ -0,0 +1,25 @@
+    - "integer"
+    - "null"
+  description: |-
+    Number of different elements in the structure as an integer.
```
```suggestion
    Number of different elements in the structure as a non-negative integer.
```

vaitkus replied:
Here I assume that 0 is allowed.

vaitkus commented on properties/src/units/current/atomistic/angstrom.yaml:
```diff
@@ -0,0 +1,16 @@
+title: "Ångström"
+symbol: "angstrom"
+$id: "https://schemas.optimade.org/units/v1.2.0/atomistic/angstrom"
+description: "The Ångström non-SI unit of length equal to $10^{-10}$ meter, where meter is defined according to the 2019 redefinition of the SI base units."
```
```suggestion
description: "The ångström non-SI unit of length equal to $10^{-10}$ meter, where meter is defined according to the 2019 redefinition of the SI base units."
```

rartino replied:
Oh, right - good catch!

vaitkus commented on schemas/src/properties/current/optimade/common/chemical_symbol.yaml:
```diff
@@ -0,0 +1,22 @@
+  - ["Li", "X"]
+type:
+  - "string"
+enum: ["X", "H", "He", "Li", "Be", "B", "C", "N", "O", "F", "Ne", "Na", "Mg", "Al", "Si", "P", "S", "Cl", "Ar", "K", "Ca", "Sc", "Ti", "V", "Cr", "Mn", "Fe", "Co", "Ni", "Cu", "Zn", "Ga", "Ge", "As", "Se", "Br", "Kr", "Rb", "Sr", "Y", "Zr", "Nb", "Mo", "Tc", "Ru", "Rh", "Pd", "Ag", "Cd", "In", "Sn", "Sb", "Te", "I", "Xe", "Cs", "Ba", "La", "Ce", "Pr", "Nd", "Pm", "Sm", "Eu", "Gd", "Tb", "Dy", "Ho ...
```
Elements 113-118 have proper symbols for some time now ("Nh", "Fl", "Mc", "Lv", "Ts", "Og" instead of "Uut", "Uuq", "Uup", "Uuh", "Uus", "Uuo") respectively. Not sure if removing the Uu* raises any issues of backwards compatibility so I left them untouched. The same change should probably also be made for other properties that enumerate the elements.

rartino replied:
You are truly doing an amazing job "proofreading" these commits. Thank you!

ml-evs commented on schemas/src/types/current/optimade/structures.yaml:
```diff
@@ -0,0 +1,158 @@
+    support: "must"
+    sortable: true
+    query-support: "all mandatory"
+    response-default-level: "alway"
```
```suggestion
    response-default-level: "always"
```

rartino commented:
A brief note here about discussing the tricky situation with the fact that SI units change slightly over time. We concluded that we have to deal with that there is a lot of data out there where the precise unit used is not known, and that there are cases where this matter, in particularity in highly precise computer calculations. 

Furthermore, we don't want to relax the principle that field definitions define a specific unit used for the value for that field.

The current proposed solution is thus to create a set of "looser" unit definitions based on SI units to be able to define, e.g., "this field is in *one* of the historical or current definitions of electron volt" with the consequence that values in this field generally should not be interpreted as precise to a higher degree than the precision of the 1960 SI units. However, when using metadata fields (see #462) we can use them to *per-entry* specify a more precise unit. So, if you know your code is delivering post-2019 SI-connected electron volts (which is EXACTLY 1.602 176 634×10⁻¹⁹ J) -- rather than the CODATA2014-value based SI electron volt (~ 1.602 176 6208(98)×10⁻¹⁹ J), or, for that matter, the 1.602 190 1(81) 10⁻¹⁹ J CODATA1969 electron volt -- you can specify this in the metadata.

merkys commented:
@rartino I remember you mentioning there will be a method to validate the property definitions using some sort of meta-schema. It would be great to develop/have the validator alongside to catch basic semantic issues.

rartino commented:
The present version validates all property / unit / etc definitions against the respective meta schemas under schemas/src/meta

vaitkus commented on schemas/src/units/current/si/general/minute.yaml:
```diff
@@ -0,0 +1,38 @@
+
+  The minute appear in the International System of Units (SI), 1th ed. (1970) defined as "1 min = 60 s".
+
+  The International System of Units (SI), 1th ed. (1970) categorizes the unit as "Units in use with the International Sytem."
```
```suggestion
  The International System of Units (SI), 1th ed. (1970) categorizes the unit as "Units in use with the International System."
```

vaitkus replied:
The word "System" is misspelt as "Sytem" in multiple files. It will probably be faster to grep for those instances then for me to mark them here manually.

rartino commented:
@JPBergsma , @giovannipizzi Do you know where I can find the dimension names that you drafted during the workshop? I have understood it such that I am expected to add them to the property definitions in this PR.

rartino commented:
Alright, I think this is finally getting ready to exit the [WIP] status:

- As far as I know, every current definition is there (but as we merge e.g., #464, those need to also be added).
- I've improved meta-validation as far as I've been able, and added a few other sanity checks. 
- The definitions in the present state of this PR are live and browsable (marked as EXPERIMENTAL) via the [schemas repo](https://github.com/Materials-Consortia/schemas) and [https://schemas.optimade.org](https://schemas.optimade.org).

However, I will let the draft status of the PR remain until we've merged #460 (since those changes are necessary for the files that are added by this PR to be correct).

oke464 commented on schemas/src/defs/v1.2/units/si/general/arcsecond.yaml:
```diff
@@ -0,0 +1,45 @@
+      id: "https://schema.optimade.org/units/si/general/radian"
+  base-units-expression: "pi*rad"
+  scale:
+    denomenator: 648000
```
```suggestion
    denominator: 648000
```

oke464 commented on schemas/src/defs/v1.2/units/si/1964/temporary/curie.yaml:
```diff
@@ -0,0 +1,33 @@
+      id: "https://schema.optimade.org/units/si/1960/named/second"
+  base-units-expression: "s^-1"
+  scale:
+    nominator: 37
```
```suggestion
    numerator: 37
```

oke464 commented on schemas/src/defs/v1.2/units/si/1970/accepted/arcminute.yaml:
```diff
@@ -0,0 +1,45 @@
+      id: "https://schema.optimade.org/units/si/1960/supplementary/radian"
+  base-units-expression: "pi*rad"
+  scale:
+    denomenator: 10800
```
```suggestion
    denominator: 10800
```

oke464 commented on schemas/src/defs/v1.2/units/si/1970/accepted/arcsecond.yaml:
```diff
@@ -0,0 +1,40 @@
+      id: "https://schema.optimade.org/units/si/1960/supplementary/radian"
+  base-units-expression: "pi*rad"
+  scale:
+    denomenator: 648000
```
```suggestion
    denominator: 648000
```

oke464 commented on schemas/src/defs/v1.2/units/si/1970/accepted/degree.yaml:
```diff
@@ -0,0 +1,35 @@
+      id: "https://schema.optimade.org/units/si/2019/named/radian"
+  base-units-expression: "pi*rad"
+  scale:
+    denomenator: 180
```
```suggestion
    denominator: 180
```

oke464 commented on schemas/src/defs/v1.2/units/si/general/arcminute.yaml:
```diff
@@ -0,0 +1,50 @@
+      id: "https://schema.optimade.org/units/si/general/radian"
+  base-units-expression: "pi*rad"
+  scale:
+    denomenator: 10800
```
```suggestion
    denominator: 10800
```

oke464 commented on schemas/src/defs/v1.2/units/si/general/curie.yaml:
```diff
@@ -0,0 +1,42 @@
+      id: "https://schema.optimade.org/units/si/general/second"
+  base-units-expression: "s^-1"
+  scale:
+    nominator: 37
```
```suggestion
    numerator: 37
```

oke464 commented on schemas/src/defs/v1.2/units/si/general/degree.yaml:
```diff
@@ -0,0 +1,42 @@
+      id: "https://schema.optimade.org/units/si/general/radian"
+  base-units-expression: "pi*rad"
+  scale:
+    denomenator: 180
```
```suggestion
    denominator: 180
```

oke464 commented on schemas/src/defs/v1.2/units/si/1998/accepted/dalton.yaml:
```diff
@@ -0,0 +1,43 @@
+    base-units-expression: "kg"
+    scale:
+      value: 1.66053906660E-27
+      uncertainty: 0.00000000050E-27
```
```suggestion
      standard_uncertainty: 0.00000000050E-27
```

oke464 commented on schemas/src/defs/v1.2/units/si/general/dalton.yaml:
```diff
@@ -0,0 +1,50 @@
+    base-units-expression: "kg"
+    scale:
+      value: 1.66053906660E-27
+      uncertainty: 0.00000000050E-27
```
```suggestion
      standard_uncertainty: 0.00000000050E-27
```

rartino commented:
Quick note: I thought I was more or less finished, but with the merging of #464, this PR needs to be updated for the new symmetry properties.

ml-evs commented:
> Quick note: I thought I was more or less finished, but with the merging of #464, this PR needs to be updated for the new symmetry properties.

If you haven't already started, perhaps I can make a PR onto your branch that adds them? Would be helpful for me to get a feel for what needs to be done to add a new field

rartino commented:
@ml-evs 

> If you haven't already started, perhaps I can make a PR onto your branch that adds them? Would be helpful for me to get a feel for what needs to be done to add a new field

Yes please, that would be very useful!

Note that the latest documentation of the formats is in another branch #460, which I updated yesterday evening - so you may want to refer to `optimade.rst` *from that branch* while working. In the branch backing the current PR there is also documentation (mostly redundant with #460 `optimade.rst`) in the description fields in the meta-schemas under `schemas/src/meta/v1.2/optimade` and in the `schemas/REAMDE.md`.

rartino commented:
> One fun issue is how we keep these definitions in sync with the spec, e.g., after my suggested changes in https://github.com/Materials-Consortia/OPTIMADE/pull/486 I will now have to update both the optimade-python-tools version, and the version here. Would you say the plan going forward is to only contribute to the machine-readable versions and then we update the spec automatically? Or for a while will there still be some back-and-forth where the spec is updated and then we manually fix the machine-readable versions (in separate PRs)?

I think it was fairly universally agreed on the last meeting that once we've managed to merge the property definitions, we should just remove them out of the specification entirely and refer to the definition files + the human readable ones at schemas.optimade.org. 

How to handle things on the side of optimade-python-tools, I don't know. Is it entirely out of the question of writing some kind of translator of the property definitions into the pydantic format needed? Perhaps I can help adding that as an output format from the converter tool if you can point me at what files in optimade-python-tools we hope to generate.

rartino commented:
In a joint effort with @ml-evs I believe this PR is in sync with #464.

Based on discussions with @ml-evs, I have moved the process_schemas tool to a separate repository along with the example for how to create database-specific property definitions: 

- https://github.com/Materials-Consortia/optimade-property-tools

I've then removed it from this PR, and added the dependency as a git submodule in `dependencies/submodules/optimade-property-tools/`.

Building the schemas `make schemas` complains if the submodule isn't initialized, and outputs the command to run to do so.

The only things I am aware of blocking this PR are:

- Sorting out the 'pattern' property definition field via #488, #490 
- modifying the description of the `pattern` field of `schemas/src/meta/v1.2/optimade/property_definition.yaml` to match.
- inserting @ml-evs property definition-compliant regex for validation of symops from #488 into the `pattern` field in `schemas/src/defs/v1.2/properties/optimade/common/space_group_symmetry_operation_xyz.yaml`.

vaitkus commented on schemas/src/defs/v1.2/units/si/1970/temporary/roentgen.yaml:
```diff
@@ -0,0 +1,34 @@
+      id: "https://schemas.optimade.org/defs/v1.2/units/si/1967/named/coulomb"
+    - symbol: "kg"
+      id: "https://schemas.optimade.org/defs/v1.2/units/si/1960/base/kilogram"
+  base-units-expression: "C*kg^-1"
+  scale:
+    exponent: -4
```
Doesn't this definition also require a numerator of 2.58 in some form here?

vaitkus commented on schemas/src/defs/v1.2/units/si/1979/named/lux.yaml:
```diff
@@ -0,0 +1,40 @@
+    resource-id: "https://en.wikipedia.org/wiki/Lux"
+defining-relation:
+  base-units:
+    - symbol: "cd"
+      id: "https://schemas.optimade.org/defs/v1.2/units/si/1967/base/candela"
+    - symbol: "sr"
+      id: "https://schemas.optimade.org/defs/v1.2/units/si/1960/supplementary/steradian"
+    - symbol: "m"
+      id: "https://schemas.optimade.org/defs/v1.2/units/si/1960/base/metre"
```
Shouldn't these refer to the redefined units? Currently, the references are seemingly the same as in `1967/named/lux`.

rartino replied:
Indeed; I also found the same thing at `./1983/named/lumen.yaml`.

vaitkus reviewed (commented):
I haven't gone through all of the unit definitions yet, but hopefully this is useful.

rartino commented on schemas/src/defs/v1.2/prefixes/si/exbi.yaml:
```diff
@@ -0,0 +1,23 @@
+  base-units-expression: ""
+  scale:
+    base: 2
+    numerator: 60
```
Indeed, good that you caught this.

I now also see that `base`, while being in the meta-schema for unit definitions, is not yet documented in optimade.rst.

rartino commented on schemas/src/defs/v1.2/prefixes/si/pebi.yaml:
```diff
@@ -0,0 +1,23 @@
+$$schema: "https://schemas.optimade.org/meta/v1.2/optimade/prefix_definition"
+$id: "https://schemas.optimade.org/defs/v1.2/prefixes/si/pebi"
+title: "pebi"
+symbol: "Ei"
```
Apparently the binary prefixes were added a bit too hastily; good catch.

rartino commented on schemas/src/defs/v1.2/units/si/1967/named/weber.yaml:
```diff
@@ -0,0 +1,42 @@
+      id: "https://schemas.optimade.org/defs/v1.2/units/si/1960/base/metre"
+    - symbol: "s"
+      id: "https://schemas.optimade.org/defs/v1.2/units/si/1967/base/second"
+  base-units-expression: "ampere^-1*kg*m^2*s^-3"
```
This is dangerous. A lot of things are caught by the process_schema validator - this is not. Perhaps it should be added.

rartino commented on schemas/src/defs/v1.2/units/si/1960/supplementary/steradian.yaml:
```diff
@@ -0,0 +1,36 @@
+
+  The radian is defined at the 9th CGPM Meeting in 1948 as: "The unit of solid angle is the steradian. The steradian is the solid angle with its vertex at the center of a sphere with a radius of one meter, which cuts out an area of one square meter on the surface of that sphere."
+
+  The International System of Units (SI), 9th Edition give a rephrased but equivalent definition: "The solid angle, expressed in steradian, corresponds to the ratio between an area \(A\) of the surface of a sphere of radius \(r\) and the squared radius, \(\Omega = A/r^2\) sr. One steradian corresponds to the solid angle for which \(A = r^2\), thus 1 sr = 1."
```
\( ... \) means render as MathJax, so that indicates the cursive variable r. With that in mind, I think everything looks correctly, but I may have gone blind by staring at these for too long :)

vaitkus replied:
Sorry, did not realise this is MathJax. Looks good then.

rartino commented on schemas/src/defs/v1.2/units/si/1970/temporary/angstrom.yaml:
```diff
@@ -0,0 +1,32 @@
+defining-relation:
+  base-units:
+    - symbol: "m"
+      id: "https://schemas.optimade.org/defs/v1.2/units/si/1983/base/metre"
```
I *think* I've fixed all these now; but it would be nice if you can double check the places you've found.

rartino replied:
This remark allowed me to realize and fix a bug in the id sanity checker. Previously it did not correctly sanity-check ids in `approximate-relations/base-units`. This caught a couple of new errors.

I indeed think referring to future constants/units make sense for `approximate-relations` when the relationship is experimentally determined, to represent these relationships as accurately as possible.

I've tried searching for the relevant cases the following way (which I'm not 100% certain catches all cases):

```
> find schemas/src -name "*.yaml" | xargs grep "id:" | grep -oP '/\d\d\d\d/.*/\d\d\d\d/' | awk -F/ '{if ($2<$(NF-1)) {print($0)}}'

1969/electromagnetic/bohrmagneton.yaml:        id: "https://schemas.optimade.org/defs/v1.2/constants/codata/2018/
/1970/accepted/atomicmassunit.yaml:      id: "https://schemas.optimade.org/defs/v1.2/units/si/2019/
/1970/accepted/electronvolt.yaml:        id: "https://schemas.optimade.org/defs/v1.2/constants/codata/2018/
/1998/accepted/dalton.yaml:      id: "https://schemas.optimade.org/defs/v1.2/units/si/2019/
/1983/accepted/electronvolt.yaml:        id: "https://schemas.optimade.org/defs/v1.2/constants/codata/2018/
/2018/electromagnetic/bohrmagneton.yaml:    id: "https://schemas.optimade.org/defs/v1.2/units/si/2019/
/2018/electromagnetic/bohrmagneton.yaml:    id: "https://schemas.optimade.org/defs/v1.2/units/si/2019/
/2018/electromagnetic/elementarycharge.yaml:      id: "https://schemas.optimade.org/defs/v1.2/units/si/2019/
```
Going through these 8 cases manually, the forward references look ok to me. The weirdest ones are the 2018 -> 2019 references, which are there because the CODATA 2018 paper defines constants in 2018 while referring to the new SI unit definitions that only took effect in 2019.

vaitkus replied:
Great, thanks.

vaitkus commented on schemas/src/defs/v1.2/units/si/1983/temporary/are.yaml:
```diff
@@ -0,0 +1,34 @@
+description: |-
+  A unit of area equal to 100 square meters (10² m²), with meter defined according to the 1983 SI units.
+
+  The are appear in the International System of Units (SI), 1st ed. (1970) defined as "1 a = 1 dam² = 10² m²" with the footnote "This unit and its symbol were adopted by the International Committee in 1879 (Proces-Verbaux C.I.P.M., 1879, p. 41).
```
Shouldn't here and in other definitions the unit be followed by a singular verb, as in "The are [unit] appears"?

rartino replied:
Not sure the "unit" really is needed, but it also doesn't hurt, so I've search-replace-inserted it everywhere.

vaitkus commented on schemas/src/defs/v1.2/units/si/general/are.yaml:
```diff
@@ -0,0 +1,40 @@
+description: |-
+  A unit of area equal to 100 square meters (10² m²) using the current, or one of the historical, definitions of the SI units.
+
+  The are appear in the International System of Units (SI), 1st ed. (1970) defined as "1 a = 1 dam² = 10² m²" with the footnote "This unit and its symbol were adopted by the International Committee in 1879 (Proces-Verbaux C.I.P.M., 1879, p. 41).
```
```suggestion
  The are appear in the International System of Units (SI), 1st ed. (1970) defined as "1 a = 1 dam² = 10² m²" with the footnote "This unit and its symbol were adopted by the International Committee in 1879 (Proces-Verbaux C.I.P.M., 1879, p. 41)."
```

vaitkus commented on schemas/src/defs/v1.2/units/si/general/hectare.yaml:
```diff
@@ -0,0 +1,41 @@
+description: |-
+  A unit of area equal to 1000 square meters (1 hm²), using the current, or one of the historical, definitions of the SI units.
+
+  The hectare appear in the International System of Units (SI), 1st ed. (1970) defined as "1 ha = 1 hm² = 10⁴ m²" with the footnote "This unit and its symbol were adopted by the International Committee in 1879 (Proces-Verbaux C.I.P.M., 1879, p. 41).
```
```suggestion
  The hectare appear in the International System of Units (SI), 1st ed. (1970) defined as "1 ha = 1 hm² = 10⁴ m²" with the footnote "This unit and its symbol were adopted by the International Committee in 1879 (Proces-Verbaux C.I.P.M., 1879, p. 41)."
```

vaitkus commented on schemas/src/defs/v1.2/units/si/1970/temporary/are.yaml:
```diff
@@ -0,0 +1,32 @@
+description: |-
+  A unit of area equal to 100 square meters (10² m²), with meter defined according to the 1960 SI base units.
+
+  The are appear in the International System of Units (SI), 1st ed. (1970) defined as "1 a = 1 dam² = 10² m²" with the footnote "This unit and its symbol were adopted by the International Committee in 1879 (Proces-Verbaux C.I.P.M., 1879, p. 41).
```
```suggestion
  The are appear in the International System of Units (SI), 1st ed. (1970) defined as "1 a = 1 dam² = 10² m²" with the footnote "This unit and its symbol were adopted by the International Committee in 1879 (Proces-Verbaux C.I.P.M., 1879, p. 41)."
```

vaitkus commented on schemas/src/defs/v1.2/units/si/1970/temporary/hectare.yaml:
```diff
@@ -0,0 +1,31 @@
+description: |-
+  A unit of area equal to 1000 square meters (1 hm²), with meter defined according to the 1960 SI base unit.
+
+  The hectare appear in the International System of Units (SI), 1st ed. (1970) defined as "1 ha = 1 hm² = 10⁴ m²" with the footnote "This unit and its symbol were adopted by the International Committee in 1879 (Proces-Verbaux C.I.P.M., 1879, p. 41).
```
```suggestion
  The hectare appear in the International System of Units (SI), 1st ed. (1970) defined as "1 ha = 1 hm² = 10⁴ m²" with the footnote "This unit and its symbol were adopted by the International Committee in 1879 (Proces-Verbaux C.I.P.M., 1879, p. 41)."
```

vaitkus commented on schemas/src/defs/v1.2/units/si/1983/named/farad.yaml:
```diff
@@ -0,0 +1,42 @@
+
+  "The farad is the capacitance of an electric capacitor between whose plates an electric potential difference of one volt appears when it is charged with an amount of electricity equal to one coulomb." [9th CGPM meeting in 1946]
+
+  The farad was defined at the 9th CGPM Meeting in 1946, included in SI at the 11th CGPM meeting in 1960, resolution 12m implicitly redefined via the redefinition of the second at the 13th CGPM Meeting in 1967, resolution 1, and again implicity redefined via the redefinition of the metre at the 17th CGPM meeting (1983), resolution 1.
```
```suggestion
  The farad was defined at the 9th CGPM Meeting in 1946, included in SI at the 11th CGPM meeting in 1960, resolution 12, implicitly redefined via the redefinition of the second at the 13th CGPM Meeting in 1967, resolution 1, and again implicity redefined via the redefinition of the metre at the 17th CGPM meeting (1983), resolution 1.
```

vaitkus replied:
Not a 100 % sure that `resolution 12m` is incorrect, please check.

rartino replied:
No reason for an m there. It appears listed here just as resolution 12: https://www.bipm.org/en/committees/cg/cgpm/11-1960/resolution-12

vaitkus commented on schemas/src/defs/v1.2/units/si/general/dalton.yaml:
```diff
@@ -0,0 +1,51 @@
+
+  This definition makes the dalton equivalent to the atomic mass unit defined in The International System of Units (SI), 1st ed. (1970) (https://schemas.optimade.org/defs/v1.2/units/si/general/atomicmassunit)
+
+  In the The International System of Units (SI) 8th ed. (2006) the definition is slightly adjusted, replacing "unbound" with "free": "The dalton (Da) and the unified atomic mass unit (u) are alternative names (and symbols) for the same unit, equal to 1/12 times the mass of a free carbon 12 atom, at rest and in its ground state."
```
```suggestion
  In the International System of Units (SI) 8th ed. (2006) the definition is slightly adjusted, replacing "unbound" with "free": "The dalton (Da) and the unified atomic mass unit (u) are alternative names (and symbols) for the same unit, equal to 1/12 times the mass of a free carbon 12 atom, at rest and in its ground state."
```

rartino replied:
You are correcting a double "the The" here -- but I also see that I have routinely capitalized "The" in these phrasings, and probably should have. I'm search-replacing them all to 'the'.

vaitkus commented on schemas/src/defs/v1.2/units/si/1998/accepted/dalton.yaml:
```diff
@@ -0,0 +1,44 @@
+
+  This definition makes the dalton equivalent to the atomic mass unit defined in The International System of Units (SI), 1st ed. (1970) (https://schemas.optimade.org/defs/v1.2/units/si/1970/accepted/atomicmassunit)
+
+  In the The International System of Units (SI) 8th ed. (2006) the definition is slightly adjusted, replacing "unbound" with "free": "The dalton (Da) and the unified atomic mass unit (u) are alternative names (and symbols) for the same unit, equal to 1/12 times the mass of a free carbon 12 atom, at rest and in its ground state."
```
```suggestion
  In the International System of Units (SI) 8th ed. (2006) the definition is slightly adjusted, replacing "unbound" with "free": "The dalton (Da) and the unified atomic mass unit (u) are alternative names (and symbols) for the same unit, equal to 1/12 times the mass of a free carbon 12 atom, at rest and in its ground state."
```

vaitkus commented on schemas/src/defs/v1.2/units/si/general/astronomicalunit.yaml:
```diff
@@ -0,0 +1,46 @@
+
+  The current definition in The International System of Units (SI), 9th ed. (2019) is: "1 au = 149 597 870 700 m" with the footnote "As decided at the XXVIII General Assembly of the International Astronomical Union (Resolution B2, 2012)."
+
+  The prior definition in The International System of Units (SI), 1st ed. (1970) appear in a footnote: "The astronomical unit of distance is the length of the unperturbed circular orbit about the Sun of a body of infinitesimal mass in motion around the Sun with an angular sidereal velocity of 0.017202098950 radians per ephemeris day of 86,400 seconds. In the astronomical constants system of the I ...
```
```suggestion
  The prior definition in The International System of Units (SI), 1st ed. (1970) appear in a footnote: "The astronomical unit of distance is the length of the unperturbed circular orbit about the Sun of a body of infinitesimal mass in motion around the Sun with an angular sidereal velocity of 0.017202098950 radians per ephemeris day of 86,400 seconds. In the System of Astronomical Constants of the International Astronomical Union, the adopted value is: 1 AU = 149,600 x 10⁶ m."
```

vaitkus replied:
Not sure if this is something we care about, but there seems to be several editions of those the IAU System of Astronomical Constants as well.

rartino replied:
In the quote appearing in French in the SI Brochure eds 1-3 (and, as far as I know, never given an official English translation (?)) they do not indicate anything about which edition(s). Given that we refer to SI:s definition, which quotes this value, I don't think we *need* to track down the origin further. (If you want to do it to clarify this quote, I'd be all for including it, though.)

vaitkus commented on schemas/src/defs/v1.2/units/si/1970/accepted/astronomicalunit.yaml:
```diff
@@ -0,0 +1,43 @@
+description: |-
+  A length representing the mean distance between the Earth and the Sun defined via the circular orbit radius for a body with a specific angular velocity.
+
+  The International System of Units (SI), 1st ed. (1970) defines the unit in a footnote as "The astronomical unit of distance is the length of the unperturbed circular orbit about the Sun of a body of infinitesimal mass in motion around the Sun with an angular sidereal velocity of 0.017202098950 radians per ephemeris day of 86,400 seconds. In the astronomical constants system of the International ...
```
```suggestion
  The International System of Units (SI), 1st ed. (1970) defines the unit in a footnote as "The astronomical unit of distance is the length of the unperturbed circular orbit about the Sun of a body of infinitesimal mass in motion around the Sun with an angular sidereal velocity of 0.017202098950 radians per ephemeris day of 86,400 seconds. In the System of Astronomical Constants of the International Astronomical Union, the adopted value is: 1 AU = 149,600 x 10⁶ m."
```

vaitkus commented on schemas/src/defs/v1.2/unitsystems/si/si_general.yaml:
```diff
@@ -0,0 +1,190 @@
+$id: "https://schemas.optimade.org/defs/v1.2/unitsystems/si/si_general"
+title: "SI general"
+description: |-
+  The SI unit system, using of current or historical definitions of the SI units.
```
```suggestion
  The SI unit system, using current or historical definitions of the SI units.
```

vaitkus commented on schemas/src/defs/v1.2/unitsystems/si/si_2019.yaml:
```diff
@@ -0,0 +1,118 @@
+  The International System of Units (SI) is established and maintained by the General Conference on Weights and Measures (CGPM).
+  In the redefinition of the SI unit system in 2019 the magnitudes of all SI units are defined from declaring exact numerical values for seven defining constants when expressed in terms of their SI units.
+
+  The SI 2019 unit system represents the original publication of the 9th edition of the The International System of Units in march 2019 (marked v1.06).
```
```suggestion
  The SI 2019 unit system represents the original publication of the 9th edition of the International System of Units in March 2019 (marked v1.06).
```

vaitkus commented on schemas/src/meta/v1.2/optimade/property_definition.yaml:
```diff
@@ -0,0 +1,432 @@
+        description: |-
+          A human-readable multi-line description that explains the purpose, requirements, and conventions of the defined property.
+          The format SHOULD be a one-line description, followed by a new paragraph (two newlines), followed by a more detailed description of all the requirements and conventions of the defined property.
+          Formatting in the text SHOULD use Markdown in the CommonMark v0.3 format format, with mathematical expressions written to render correctly with the LaTeX mode of Mathjax 3.2 <https://docs.mathjax.org/en/v3.2-latest/>.
```
```suggestion
          Formatting in the text SHOULD use Markdown in the CommonMark v0.3 format, with mathematical expressions written to render correctly with the LaTeX mode of Mathjax 3.2 <https://docs.mathjax.org/en/v3.2-latest/>.
```

vaitkus commented on schemas/src/meta/v1.2/optimade/property_definition.yaml:
```diff
@@ -0,0 +1,432 @@
+          names:
+            description: |-
+              A list of names of the dimensions of the underlying one or multi-dimensionsional data represented as mutiple levels of lists.
+              The order is that the the first name applies to the outermost list, the next name to the lists embedded in that list, etc.
```
```suggestion
              The order is that the first name applies to the outermost list, the next name to the lists embedded in that list, etc.
```

vaitkus commented on schemas/src/meta/v1.2/optimade/property_definition.yaml:
```diff
@@ -0,0 +1,432 @@
+          sizes:
+            description: |-
+              A list of fixed length requirements on the underlying one or multi-dimensionsional data represented as mutiple levels of lists.
+              The order is that the the first name applies to the outermost list, the next name to the lists embedded in that list, etc.
```
```suggestion
              The order is that the first name applies to the outermost list, the next name to the lists embedded in that list, etc.
```

vaitkus commented on schemas/src/meta/v1.2/optimade/property_definition.yaml:
```diff
@@ -0,0 +1,432 @@
+          Comments pertaining to the Property Definition that are relevant to end users should go into the field description.
+          Formatting in the text SHOULD use Markdown using the format described in the definition of the description field.
+
+          This fields is an annotation rather than an integral part of the Property Definition.
```
```suggestion
          This field is an annotation rather than an integral part of the Property Definition.
```

vaitkus commented on schemas/src/meta/v1.2/optimade/property_definition.yaml:
```diff
@@ -0,0 +1,432 @@
+           If TRUE, implementations SHOULD not use the defined property, and it MAY be removed in the future.
+           If FALSE, the defined property is not deprecated. The field not being present means FALSE.
+
+           This fields is an annotation rather than an integral part of the Property Definition.
```
```suggestion
           This field is an annotation rather than an integral part of the Property Definition.
```

rartino commented on schemas/src/defs/v1.2/unitsystems/si/si_1973.yaml:
```diff
@@ -0,0 +1,94 @@
+$id: "https://schemas.optimade.org/defs/v1.2/unitsystems/si/si_1973"
+title: "SI 1973"
+description: |-
+  The SI unit system as represented in the 3rd edition of The International System of Units.
```
Actually in two places in that file (!) Thanks for catching this!

rartino commented on schemas/src/defs/v1.2/units/si/1970/temporary/hectare.yaml:
```diff
@@ -0,0 +1,31 @@
+$$schema: "https://schemas.optimade.org/meta/v1.2/optimade/physical_unit_definition"
+$id: "https://schemas.optimade.org/defs/v1.2/units/si/1970/temporary/hectare"
+title: "hectare"
+symbol: "he"
+display-symbol: "he"
```
Indeed; it is there in the description text which quotes the SI document saying "ha"...

vaitkus commented on schemas/src/defs/v1.2/units/si/general/electronvolt.yaml:
```diff
@@ -0,0 +1,39 @@
+description: |-
+  A unit of energy that representing the kinetic energy acquired by an electron as it accelerates through a 1 volt potential difference in a vacuum using the current, or one of the historical, definitions given in the editions of the International System of Units (SI).
+
+  The electronvolt appear in the International System of Units (SI), 1st ed. (1970) defined as "1 electronvolt is the energy acquired by an electron after traversing a potential difference of 1 V in a vacuum; 1 eV = 1.60219×10⁻¹⁹ J approximately."
```
```suggestion
  The electronvolt unit appears in the International System of Units (SI), 1st ed. (1970) defined as "1 electronvolt is the energy acquired by an electron after traversing a potential difference of 1 V in a vacuum; 1 eV = 1.60219×10⁻¹⁹ J approximately."
```

vaitkus commented on schemas/src/defs/v1.2/units/si/2019/accepted/electronvolt.yaml:
```diff
@@ -0,0 +1,34 @@
+description: |-
+  A unit of energy equal to the amount of kinetic energy acquired by an electron as it accelerates through a 1 volt potential difference in a vacuum using the 2019 SI units.
+
+  The electronvolt appear in the International System of Units (SI), 1st ed. (1970) defined as "1 electronvolt is the energy acquired by an electron after traversing a potential difference of 1 V in a vacuum; 1 eV = 1.60219×10⁻¹⁹ J approximately."
```
```suggestion
  The electronvolt unit appears in the International System of Units (SI), 1st ed. (1970) defined as "1 electronvolt is the energy acquired by an electron after traversing a potential difference of 1 V in a vacuum; 1 eV = 1.60219×10⁻¹⁹ J approximately."
```

vaitkus commented on schemas/src/defs/v1.2/units/si/1970/accepted/astronomicalunit.yaml:
```diff
@@ -0,0 +1,43 @@
+  - The International System of Units (SI), 4th ed. (1981) omits the unit.
+  - The International System of Units (SI), 7th ed. (1998) reintroduces the unit in the category "Non-SI units accepted for use with the International System, whose values in SI units are obtained experimentally." with the experimental relationship "1 au = 1.49597870691(30)×10¹¹ m" with the footnote: "The astronomical unit is a unit of length approximatively equal to the mean Earth-Sun distance.  ...
+  - The International System of Units (SI), 8th ed. (2006) notes the experimental relationship "1 au = 1.49597870691(6)×10¹¹ m" and changes the footnote to: "The astronomical unit is approximately equal to the mean Earth-Sun distance. It is the radius of an unperturbed circular Newtonian orbit about the Sun of a particle having infinitesimal mass, moving with a mean motion of 0.01720209895 radian ...
+  - The International System of Units (SI), 9th ed. (2019) redefines the unit to have an exact relationship to the meter with the footnote: "This definition is referenced in a footnote as: "As decided at the XXVIII General Assembly of the International Astronomical Union (Resolution B2, 2012)."
```
```suggestion
  - The International System of Units (SI), 9th ed. (2019) redefines the unit to have an exact relationship to the meter with the footnote "As decided at the XXVIII General Assembly of the International Astronomical Union (Resolution B2, 2012)."
```

vaitkus commented on schemas/src/defs/v1.2/units/si/1970/accepted/parsec.yaml:
```diff
@@ -0,0 +1,30 @@
+description: |-
+  The distance at which 1 astronomical unit subtends an angle of 1 arcsecond using the definition of the astronomical unit and arcsecond appearing in the International System of Units (SI), 1st ed. (1970)
+
+  The parsec unit appears in the International System of Units (SI), 1st ed. (1970) defined in a footnote as "1 parsec is the distance at which 1 astronomical unit subtends an angle of 1 arcsecond; therefore, approximately, 1 pc = 206 265 AU = 30 857 x 10¹² m.
```
```suggestion
  The parsec unit appears in the International System of Units (SI), 1st ed. (1970) defined in a footnote as "1 parsec is the distance at which 1 astronomical unit subtends an angle of 1 arcsecond; therefore, approximately, 1 pc = 206 265 AU = 30 857 x 10¹² m."
```

vaitkus commented on schemas/src/defs/v1.2/units/si/2019/named/farad.yaml:
```diff
@@ -0,0 +1,42 @@
+
+  "The farad is the capacitance of an electric capacitor between whose plates an electric potential difference of one volt appears when it is charged with an amount of electricity equal to one coulomb." [9th CGPM meeting in 1946]
+
+  The farad was defined at the 9th CGPM Meeting in 1946, included in SI at the 11th CGPM meeting in 1960, resolution 12 and implicitly redefined via the redefinitions of the second at the 13th CGPM Meeting in 1967, resolution 1, and the redefinitions of the ampere and kilogram at the 26th CGPM Meeting in 2018, resolution 1
```
```suggestion
  The farad was defined at the 9th CGPM Meeting in 1946, included in SI at the 11th CGPM meeting in 1960, resolution 12 and implicitly redefined via the redefinitions of the second at the 13th CGPM Meeting in 1967, resolution 1, and the redefinitions of the ampere and kilogram at the 26th CGPM Meeting in 2018, resolution 1.
```

vaitkus commented on schemas/src/defs/v1.2/units/si/general/astronomicalunit.yaml:
```diff
@@ -0,0 +1,46 @@
+
+  The current definition in the International System of Units (SI), 9th ed. (2019) is: "1 au = 149 597 870 700 m" with the footnote "As decided at the XXVIII General Assembly of the International Astronomical Union (Resolution B2, 2012)."
+
+  The prior definition in the International System of Units (SI), 1st ed. (1970) appear in a footnote: "The astronomical unit of distance is the length of the unperturbed circular orbit about the Sun of a body of infinitesimal mass in motion around the Sun with an angular sidereal velocity of 0.017202098950 radians per ephemeris day of 86,400 seconds. In the System of Astronomical Constants of th ...
```
```suggestion
  The prior definition in the International System of Units (SI), 1st ed. (1970) appears in a footnote: "The astronomical unit of distance is the length of the unperturbed circular orbit about the Sun of a body of infinitesimal mass in motion around the Sun with an angular sidereal velocity of 0.017202098950 radians per ephemeris day of 86,400 seconds. In the System of Astronomical Constants of the International Astronomical Union, the adopted value is: 1 AU = 149,600 x 10⁶ m."
```

vaitkus commented on schemas/src/defs/v1.2/units/si/general/astronomicalunit.yaml:
```diff
@@ -0,0 +1,46 @@
+  - The International System of Units (SI), 4th ed. (1981) omits the unit.
+  - The International System of Units (SI), 7th ed. (1998) reintroduces the unit in the category "Non-SI units accepted for use with the International System, whose values in SI units are obtained experimentally." with the experimental relationship "1 au = 1.49597870691(30)×10¹¹ m" with the footnote: "The astronomical unit is a unit of length approximatively equal to the mean Earth-Sun distance.  ...
+  - The International System of Units (SI), 8th ed. (2006) notes the experimental relationship "1 au = 1.49597870691(6)×10¹¹ m" and changes the footnote to: "The astronomical unit is approximately equal to the mean Earth-Sun distance. It is the radius of an unperturbed circular Newtonian orbit about the Sun of a particle having infinitesimal mass, moving with a mean motion of 0.01720209895 radian ...
+  - The International System of Units (SI), 9th ed. (2019) categorizes the astronomical unit as a "non-SI unit accepted for use with the SI units" and redefines it to have an exact relationship to the meter with the footnote: "This definition is referenced in a footnote as: "As decided at the XXVIII General Assembly of the International Astronomical Union (Resolution B2, 2012)."
```
```suggestion
  - The International System of Units (SI), 9th ed. (2019) categorizes the astronomical unit as a "non-SI unit accepted for use with the SI units" and redefines it to have an exact relationship to the meter with the footnote "As decided at the XXVIII General Assembly of the International Astronomical Union (Resolution B2, 2012)."
```

vaitkus commented on schemas/src/defs/v1.2/units/si/general/parsec.yaml:
```diff
@@ -0,0 +1,35 @@
+description: |-
+  A unit of length representing the distance at which 1 astronomical unit subtends an angle of 1 arcsecond using the current, or one of the historical, definitions of the SI units.
+
+  The parsec unit appears in the International System of Units (SI), 1st ed. (1970) defined in a footnote as "1 parsec is the distance at which 1 astronomical unit subtends an angle of 1 arcsecond; therefore, approximately, 1 pc = 206 265 AU = 30 857 x 10¹² m.
```
```suggestion
  The parsec unit appears in the International System of Units (SI), 1st ed. (1970) defined in a footnote as "1 parsec is the distance at which 1 astronomical unit subtends an angle of 1 arcsecond; therefore, approximately, 1 pc = 206 265 AU = 30 857 x 10¹² m."
```

vaitkus commented:
The `make schemas` command requires the `python3-mdx-math` dependency under Debian 12. @rartino, would it be OK if I would push a file with dependencies for Debian-12 directly to your branch (under `/dependencies/Debian-12`)?

Also, I noticed that the `dependencies` directory now contains subdirectories with a mixed purpose (directories that have lists of dependency installers for various OSes vs directories Git submodules). Not sure if there are any standard practices on the placement of Git submodules (the dependency installer placement comes from the COD).

rartino commented:
@vaitkus  Thanks for all the rigorous review work!

> The make schemas command requires the python3-mdx-math dependency under Debian 12. @rartino, would it be OK if I would push a file with dependencies for Debian-12 directly to your branch (under /dependencies/Debian-12)?

Sure, go ahead; but as an alternative - if you want to give me a chance to see what you change - the more straightforward way is to do a github PR against the branch in my repo that backs this PR: https://github.com/rartino/OPTIMADE/tree/add-property-definitons-3

> Also, I noticed that the dependencies directory now contains subdirectories with a mixed purpose (directories that have lists of dependency installers for various OSes vs directories Git submodules). Not sure if there are any standard practices on the placement of Git submodules (the dependency installer placement comes from the COD).

Well, on a higher level, these purposes seems very similar to me; they handle dependencies - some via shell commands that install them -  and others git via submodules. But I'm happy to re-organize this according to suggestions. Maybe we can move the installers down one level with a suitable naming? `dependencies/installers`?

rartino commented:
> WARNING: JSON Schema Python library is not aware of the Draft202012 standard. It is probably too old. Will validate using default validator.

This warning message means exactly what it says. Surprisingly recent versions of the jsonschema Python package are needed to truly validate against the correct draft standard. But, I've largely managed to validate the schemas successfully using older jsonschema packages despite this warning. But if that does not work, the workaround if the OS repo version is too old is to install a newer one via pip (or conda).

> Validation of data failed. Error details:
> - TypeError: unhashable type: 'collections.OrderedDict' (/home/antanas/src/OPTIMADE/dependencies/submodules/optimade-property-tools/src/process_schemas/process_schemas.py line 1471 triggered in: /usr/lib/python3/dist-packages/jsonschema/validators.py at line 841).

This looks like a bug in the script, which strangely I didn't see yet (and apparently nor the schema audit CI). I think your quote of the failing command misses the first part; can you make sure the whole failing command is shown?

rartino commented:
@vaitkus 

> This looks like a bug in the script, which strangely I didn't see yet (and apparently nor the schema audit CI). I think your quote of the failing command misses the first part; can you make sure the whole failing command is shown?

Actually, after digging a bit further I suspect it may still be a jsonschema package version issue. Nevertheless, it would be helpful if you can:

1. Re-run the failing command with the debug flag `-d` to get a full Python traceback
2. Open the origin of the error inside the jsonschema package `/usr/lib/python3/dist-packages/jsonschema/validators.py` at line 841 and provide some context of which of what it is trying to do there, i.e., can we figure out which of our input parameters to the jsonschema 'validate' function it is trying to insert somewhere where it needs a hash:able type.

rartino commented:
@vaitkus Please try now; pull the latest version of this branch, and update optimade-property-tools to the matching version (`git submodule update --init dependencies/submodules/optimade-property-tools` or just pull from develop).

Turns out the various releases of JSON Schema and the python-jsonschema does things a bit differently. I've tried to both improve the validation implementation and restructure the meta-schemas to increase compatibility across different versions.

vaitkus commented:
@rartino, thank you for the changes. The code now works both on Debian 11 and Debian 12. I also diff'ed the produced output and the only difference between the two systems is in the generated HTML files (the ones without the extensions). Even with the mentioned differences the files seem to produce the same visual result.

The command on Debian 10 still fails, although with a different message:
```
Traceback (most recent call last):
  File "dependencies/submodules/optimade-property-tools/bin/process_schemas", line 9, in <module>
    from process_schemas import main
  File "/home/antanas/src/rartino/OPTIMADE/dependencies/submodules/optimade-property-tools/src/process_schemas/__init__.py", line 1, in <module>
    from .process_schemas import main
  File "/home/antanas/src/rartino/OPTIMADE/dependencies/submodules/optimade-property-tools/src/process_schemas/process_schemas.py", line 44, in <module>
    from importlib import metadata
ImportError: cannot import name 'metadata' from 'importlib' (/usr/lib/python3.7/importlib/__init__.py)
``` 
Please let me know if you need any more details on this.

rartino commented:
@vaitkus 

> from importlib import metadata
> ImportError: cannot import name 'metadata' from 'importlib' (/usr/lib/python3.7/importlib/__init__.py)

This is a Python version incompatibility for < 3.8. I've added a fix, but it may need the additional dependency python3-importlib-metadata from the Debian-10 backports repo. That dependency isn't needed for Debian-11 and 12 (but I also don't think it hurts to install it).

vaitkus commented:
@rartino, ok, I installed the dependency, but now the earlier errors are back:
```
WARNING: Could not determine jsonschema version
WARNING: JSON Schema Python library is not aware of the Draft202012 standard. It is probably too old (< 4). Will validate using default validator.
Validation of data failed. Error details:
- RefResolutionError: Validation: attempt to fetch remote schema over the internet blocked: meta/core (/home/antanas/src/rartino/OPTIMADE/dependencies/submodules/optimade-property-tools/src/process_schemas/process_schemas.py line 1493 triggered in: /usr/lib/python3/dist-packages/jsonschema/validators.py at line 385).
```
Did you get it working on your Debian 10 machine? If not, I think we might just add a caveat that this make rule will not work on Debian 10 with the packages from the default repo. Well, of course unless you want to figure this out -- I would simply hate to waste your time on this if you do not see the use for it.

rartino commented:
@vaitkus 

> Did you get it working on your Debian 10 machine? 

I just looked up versions for 'buster' using https://packages.debian.org/index and tried to recreate a similar enough environment using conda. I guess I should move over to docker and a proper Debian 10 to see what is causing the issues.

The message: "WARNING: Could not determine jsonschema version" does make it sound as if the importlib.meta-thing still doesn't work for some reason.

vaitkus commented on optimade.rst:
```diff
@@ -2037,39 +2050,61 @@ The format of Property Definitions defined below allows nesting inner Property D
 
-  - :field:`property-format`: String.
-    Specifies the minor version of the property definition format used.
+  .. _definition of the property-format field:
```
```suggestion
.. _definition of the property-format field:
```

vaitkus commented on optimade.rst:
```diff
@@ -2256,18 +2369,13 @@ Depending on what string the :field:`type` is equal to, or contains as first ele
     If :val:`FALSE`, this field sets no limitation on the defined property.
 
+  Furthermore, despite being defined in the JSON Schema standard, the fields :field:`minItems` and :field:`maxItems` MUST NOT be used to indicate limits of the number of items of a list.
+  See the `definition of the x-optimade-dimension field`_ for more information.
```
```suggestion
  See the `definition of the x-optimade-dimensions field`_ for more information.
```

rartino commented:
Alright, I've now removed the 'draft' designation. I am not aware of any remaining issues with any of the definitions. The browsable versions at https://schemas.optimade.org/releases/v1.2.0/ are up to date. Please approve :)

It seems unlikely a changset of this size contains zero errors. Nevertheless, I think we have done what we can by setting up schema validation via meta-schemas and sanity checks that are run for all definition files when running `make schemas`. We've also tried to design versioning so that future point releases can address any issues we find, even in the older superseded definitions.

(Pinging @ml-evs as he is already in status 'review pending' so I cannot send an automated request for review)

vaitkus commented on README.md:
```diff
@@ -34,9 +34,9 @@ If this is a pre-release, implementations are also encouraged to support the lat
 
 The OPTIMADE standard refers to a specific version of the `definitions.units` database included with the source distribution of GNU Units.
-This file is included in the OPTIMADE repository under the subdirectory [units](units).
+This file is included in the OPTIMADE repository under the subdirectory [external/GNU_Units](external/GNU_Units).
```
The subdirectory is currently named `GNU_units`, not `GNU_Units`, thus all the links in this file are broken. Not sure whether the directory name or the links are incorrect.

rartino replied:
Since GNU is capitalized, I think it makes the most sense to also capitalize Units, so I did that. I did a "find" to check for other occurrences of "GNU_units" which only pointed at the .pre-commit-config.yaml (which I also fixed).

vaitkus commented on GNUmakefile:
```diff
@@ -19,6 +19,26 @@
+#        - schemas_html_pretty=true applies OPTIMADE styling to the html output
+#        - schemas_html_ext=true generates html files with .html extensions also
+#              for files meant to be served without extensions, which is useful
+#              for hosting, e.g., on github that automatically redirects URLs without extensions.
```
```suggestion
#              for hosting, e.g., on GitHub that automatically redirects URLs without extensions.
```

vaitkus reviewed (approved):
Ok, I did not make a thorough review this time, but I did notice a small issue with the mismatching directory names. I you want to avoid re-triggering the review for everyone, please just mark those conversations as resolved and then preferably fix the broken links after the merge to the `develop` branch.

vaitkus commented:
> The browsable versions at https://schemas.optimade.org/releases/v1.2.0/ are up to date.

@rartino This specific website might be a bit out of date since the parenthesis near "Bohr magneton" and "Elementary charge" still identify them as units, i.e.  `Bohr magneton (unit)`. Pages of the specific definitions, e.g. https://schemas.optimade.org/releases/v1.2.0/v1.2/constants/codata/2018/electromagnetic/bohrmagneton, do not have this issue.

rartino commented:
>> The browsable versions at https://schemas.optimade.org/releases/v1.2.0/ are up to date.
> @rartino This specific website might be a bit out of date since the parenthesis near "Bohr magneton" and "Elementary charge" still identify them as units, i.e. Bohr magneton (unit). 

@vaitkus Thanks, good catch. I put too much trust in the dependencies expressed in the Makefile and skipped the 'make clean' after that final fix. I re-did all definitions now after cleaning, and it changed precisely these four files: `defs/index.html`, `defs/index.md`, `releases/v1.2.0/index.html`, `releases/v1.2.0/index.md`. I guess I should try to update the full dependency specification of the index files in the Makefile at some point.

dwinston reviewed (approved):
> It seems unlikely a changset of this size contains zero errors. Nevertheless, I think we have done what we can by setting up schema validation via meta-schemas and sanity checks that are run for all definition files when running make schemas. We've also tried to design versioning so that future point releases can address any issues we find, even in the older superseded definitions.

I agree. Fantastic work, team.

merkys commented on external/README.md:
```diff
@@ -0,0 +1,3 @@
+This directory contains a collection of subfolders containing documents that are relevant to the OPTIMADE standard.
```
```suggestion
This directory contains a collection of subdirectories containing documents that are relevant to the OPTIMADE standard.
```
If we use "directory", I would suggest to stick to the same term everywhere.

merkys commented on external/README.md:
```diff
@@ -0,0 +1,3 @@
+This directory contains a collection of subfolders containing documents that are relevant to the OPTIMADE standard.
+Please note that the files in these subfolders are subject to separate licensing agreements that differ from those of other OPTIMADE files.
```
```suggestion
Please note that the files in these subdirectories are subject to separate licensing agreements that differ from those of other OPTIMADE files.
```

merkys commented on optimade.rst:
```diff
@@ -1419,21 +1423,20 @@ Example (note: the description strings have been wrapped for readability only):
-            "title": "Unit cell lattice vectors",
+            "title": "Lattice vectors",
+            "x-optimade-definition": {
+              "label": "lattice_vectors_optimade_structures"
```
```suggestion
              "label": "lattice_vectors_optimade_structures",
```

merkys commented on optimade.rst:
```diff
@@ -1419,21 +1423,20 @@ Example (note: the description strings have been wrapped for readability only):
-              "property-format": "1.2",
-              "unit-definitions": [
+            "x-optimade-dimensions": {
+               "names": ["dim_lattice", "dim_spatial"]
```
```suggestion
               "names": ["dim_lattice", "dim_spatial"],
```

vaitkus reviewed (approved):
Looks great overall, last few minor typo fixes.

merkys commented on optimade.rst:
```diff
@@ -2186,36 +2280,54 @@ The format described in this subsection forms a subset of the `JSON Schema Valid
+
+- :field:`$schema`: String.
+  A URL for a meta schema that describes the Property Definitions format.
+  For Property Definitions adhering to the format described in this document, it should be set to: "https://schemas.optimade.org/meta/v1.2/optimade/property_definition.json".
```
```suggestion
  For Property Definitions adhering to the format described in this document, it should be set to: :val:`https://schemas.optimade.org/meta/v1.2/optimade/property_definition.json`.
```

merkys commented on optimade.rst:
```diff
@@ -2341,32 +2456,54 @@ Clients and servers that use other units internally thus have to do unit convers
 
+- :field:`$schema`: String.
+  A URL for a meta schema that describes the Physical Unit Definitions format.
+  For Property Definitions adhering to the format described in this document, it should be set to: "https://schemas.optimade.org/meta/v1.2/optimade/physical_unit_definition.json".
```
```suggestion
  For Property Definitions adhering to the format described in this document, it should be set to: :val:`https://schemas.optimade.org/meta/v1.2/optimade/physical_unit_definition.json`.
```

merkys commented on schemas/REAMDE.md:
```diff
@@ -0,0 +1,273 @@
+- `<name>` is an identifier of lowercase Latin characters and the underscore character identifying the property.
+
+The URIs are URLs that can be retrieved to fetch a human-readable description of the definition in HTML format.
+Every URI can also be suffixed with the extension ".json" to obtain the machine-readable JSON definition file.
```
```suggestion
Every URI can also be suffixed with the extension `.json` to obtain the machine-readable JSON definition file.
```

merkys reviewed (approved):
I left some minor change suggestions, but nevertheless I approve.

coreyoses commented on schemas/src/defs/v1.2/properties/optimade/structures/assemblies.yaml:
```diff
@@ -0,0 +1,194 @@
+      type:
+        - "number"
+      description: |-
+        An float that specifies the statistical probability of a group in the description of assemblies.
```
``
```suggestion
        A float that specifies the statistical probability of a group in the description of assemblies.
```

vaitkus commented on schemas/src/defs/v1.2/prefixes/si/atto.yaml:
```diff
@@ -0,0 +1,22 @@
+symbol: "a"
+display-symbol: "a"
+$id: "https://schemas.optimade.org/defs/v1.2/prefixes/si/atto"
+description: "The atto SI prefix defined as a dimensionless multiple of 10⁻¹⁸, defined at the 12:th CGPM Meeting in 1964, resolution 8."
```
Here and in several other YAML files the `th` suffix is sometimes preceded by a colon (and sometimes not), e.g. `12:th`. Feel free to ignore the comment if this is intentional.

rartino commented on optimade.html:
```diff
@@ -0,0 +1,4930 @@
+<!DOCTYPE html>
```
Doh! This is a 'git add -A' gone wrong.

rartino commented:
Ok, to not tax people with re-review requests, I'll wait 24 hours; make in one go any fixes that have come in, and re-request reviews from everyone who already left them.

gmrigna commented:
Thanks a lot @rartino for this huge effort and thanks all others for their contributions.
Just a small comment on my side, I was going through the browsable version and I was wondering why some of the links are with a capital letter and some other not (a.o. there is both "url" and "URL").

oke464 commented on schemas/src/defs/v1.2/units/si/1967/named/degcelsius.yaml:
```diff
@@ -0,0 +1,39 @@
+    resource-id: "https://en.wikipedia.org/wiki/Celsius"
+defining-relation:
+  base-units:
+    - symbol: "degK"
```
```suggestion
    - symbol: "K"
```

oke464 commented on schemas/src/defs/v1.2/units/si/1967/named/degcelsius.yaml:
```diff
@@ -0,0 +1,39 @@
+  base-units:
+    - symbol: "degK"
+      id: "https://schemas.optimade.org/defs/v1.2/units/si/1967/base/kelvin"
+  base-units-expression: "degK"
```
```suggestion
  base-units-expression: "K"
```

oke464 reviewed (commented):
In the 1967 schema for Kelvin (https://schemas.optimade.org/defs/v1.2/units/si/1967/base/kelvin), the symbol used is "K", changed after 1960. The design allows for different symbols in defining-relation, so this is not a breaking change and we may keep this as is if we wish. 😊

rartino commented on schemas/src/defs/v1.2/units/si/1960/named/degcelsius.yaml:
```diff
@@ -0,0 +1,33 @@
+$$schema: "https://schemas.optimade.org/meta/v1.2/optimade/physical_unit_definition"
+$id: "https://schemas.optimade.org/defs/v1.2/units/si/1960/named/degcelsius"
+title: "Degree Celsius"
```
@vaitkus This one isn't so clear-cut to me. I know there are words/phrasings that cannot be capitalized even as the first word in a title, but those are peculiar exceptions - does it really apply here? Do you have a style guide or similar to point to?

rartino commented on tests/makefiles/Makelocal-swagger:
```diff
@@ -2,12 +2,12 @@
-
-validate_index_schema:
-	./tests/scripts/validate_swagger.sh ./schemas/index_openapi_schema.json
+#.PHONY: validate_schemas
+#
+#validate_schemas: validate_schema validate_index_schema
+#
+#validate_schema:
+#	./tests/scripts/validate_swagger.sh ./schemas/openapi_schema.json
+#
+#validate_index_schema:
+#	./tests/scripts/validate_swagger.sh ./schemas/index_openapi_schema.json
```
I'm going to argue that this PR does not make the OpenAPI schema situation worse. So, can we take this question as a separate issue + PR: "handle the OpenAPI schemas for 1.2.0"? Where we can agree what to do. Indeed, maybe the most obvious solution is to just stop providing complete OpenAPI schemas for now.

rartino replied:
Ok, I posted issue #502, and consider that the resolution of this conversation. I.e., lets take "fix OpenAPI schemas" into the next PR.

vaitkus commented on schemas/src/defs/v1.2/constants/codata/2018/electromagnetic/bohrmagneton.yaml:
```diff
@@ -1,6 +1,6 @@
 $$schema: "https://schemas.optimade.org/meta/v1.2/optimade/constant_definition"
 $id: "https://schemas.optimade.org/defs/v1.2/constants/codata/2018/electromagnetic/bohrmagneton"
-title: "Bohr magneton"
+title: "bohr magneton"
```
```suggestion
title: "Bohr magneton"
```

vaitkus replied:
In this case "Bohr" refers to the last name, so the capitalisation should probably be retained. Wikipedia capitalises it in the middle of sentences as well (see https://en.wikipedia.org/wiki/Bohr_magneton).

rartino replied:
Absolutely right-- I went through all titles to bring them into consistency but lapsed on this. Thanks!

vaitkus commented on schemas/src/defs/v1.2/units/codata/1969/electromagnetic/bohrmagneton.yaml:
```diff
@@ -1,6 +1,6 @@
 $$schema: "https://schemas.optimade.org/meta/v1.2/optimade/physical_unit_definition"
 $id: "https://schemas.optimade.org/defs/v1.2/units/codata/1969/electromagnetic/bohrmagneton"
-title: "Bohr magneton"
+title: "bohr magneton"
```
```suggestion
title: "Bohr magneton"
```

ml-evs commented on schemas/REAMDE.md:
```diff
@@ -0,0 +1,273 @@
+# OPTIMADE Property Definitions and other schemas
```
This file is currently called "REAMDE"

rartino commented:
@ml-evs @vaitkus @oke464 @gmrigna @coreyoses @merkys @dwinston 

Thanks for all reviews and feedback. The PR is once again in a no-open-conversations state and needs to be re-approved. The most major thing corrected was the inconsistency of the "title" field, which is now hopefully consistently fixed.

The rendered definitions at: https://schemas.optimade.org/releases/v1.2.0/ are refreshed with the current status.

vaitkus commented on schemas/src/defs/v1.2/properties/optimade/common/chemical_symbol.yaml:
```diff
@@ -0,0 +1,26 @@
+
+  **Requirements/Conventions:**
+
+  - Must be one of the following:
+
+    - a valid chemical-element symbol, or
+    - the special value `"X"` to represent a non-chemical element.
```
Not a show stopper, this can be fixed after the merge.

In the HTML page for this YAML [1], all three lines prefixed by (`- `) appear a elements of the same list while the last two elements were most likely intended to appear in a sublist. Same issue seems to appear in other pages as well.



[1] https://schemas.optimade.org/releases/v1.2.0/v1.2/properties/optimade/common/chemical_symbol

rartino replied:
Sigh, formally strict markdown renderers' requirement of *four spaces* for sublists always trip me up since most more sloppy renderers don't require it, and neither does restructuredText. Since it is only you and @merkys who have reviewed this round so far, lets try to get a fix in.

rartino replied:
So, I *think* I've fixed all these now, and uploaded the changes to the schemas repo, but please check the ones you were aware of.

vaitkus commented on schemas/src/defs/v1.2/properties/optimade/common/person.yaml:
```diff
@@ -0,0 +1,43 @@
+    type:
+      - string
+examples:
+  - {"name": "John Smith"}
```
On the website [1] these examples appear as:
```
    OrderedDict([('name', 'John Smith')])
    OrderedDict([('name', 'Arthur Dent'), ('firstname', 'Arthur'), ('lastname', 'Dent')])
```
Is that on purpose?

[1] https://schemas.optimade.org/releases/v1.2.0/v1.2/properties/optimade/common/person

rartino replied:
This is not on purpose, but should probably be filed against the https://github.com/Materials-Consortia/optimade-property-tools repo instead.

vaitkus replied:
Ok, will refile it there.

rartino replied:
It should be fixed now. I found a trick online to convert it to json before printing it. This has changed some other output, e.g., strings are now rendered as with quotation marks in the examples section. Do we like this?

vaitkus replied:
I do not mind the quotes too much, but I wonder how would values that contain double quotes be rendered. Would the double quote be escaped as `\"` thus modifying the specific given example?

One example that comes to mind is space group names (`P 3 2"`), although, such examples do not seem to be currently in use.

rartino replied:
Yes, it would be rendered as serialized JSON, so your example would be render as `"P 3 2\""`. The old rendering would have been just `P 3 2"`. However, IMO the new rendering is more universally consistent than the old one where examples of pure string types were rendered without quotation, but strings inside a dict were rendered with quotation marks based on Pythons `str()` algorithm, which switches between " and ' to avoid escapes if possible. Is it not more logical to consistently use JSON serialization for everything than to have our definition pages undocumented relay Python-isms?

vaitkus replied:
Documented is definitely better than undocumented so we should definitely go with this approach for now. If anybody feels strongly about this later, a separate issue can always be raised after the merge of this PR.

I would probably still go with verbatim text without any surrounding quotes in such situations since the examples are wrapped in the HTML `code` tags. We would also allow to avoid some inconsistencies that arise when examples are also provided in the description text. See for example the Hall space group [1] page which contains two sections -- `Explained examples` and `Examples` -- one of which is quoted and one is not. 

[1] https://schemas.optimade.org/releases/v1.2.0/v1.2/properties/optimade/structures/space_group_symbol_hall

ml-evs commented on schemas/README.md:
````diff
@@ -0,0 +1,273 @@
+  rsync -av --no-t --checksum schemas/output/ /path/to/repo/for/schemas/
+  ```
+
+- Check that the changes look resonable also from the perspective of git:
````
```suggestion
- Check that the changes look reasonable also from the perspective of git:
```

ml-evs commented on optimade.rst:
```diff
@@ -2074,39 +2086,61 @@ The format of Property Definitions defined below allows nesting inner Property D
 To make a property definition expressible in any output format, the fields of the property definition below are specified using OPTIMADE data types.
 When a property definition is communicated using a specific data format (e.g., JSON), the property definition is implemented in that data format by mapping the OPTIMADE data types into the corresponding data types for that output format.
 
+Clients are meant to be able to rely on that properties with the same :field:`$id` fields represents equivalently defined properties.
+Hence, when a Property Definition that has been published previously is updated, it is of major importance to decide if the update merely amend, annotate, or clarify the definition in a way that leaves it functionally the same and thus can retain the :field:`$id`, or whether the property is redefined.
```
```suggestion
Clients are meant to be able to rely on the fact that properties with the same :field:`$id` fields represents equivalently defined properties.
Hence, when a Property Definition that has been published previously is updated, it is of major importance to decide if the updates merely amend, annotate, or clarify the definition in a way that leaves it functionally the same and thus can retain the :field:`$id`, or whether the property is redefined.
```

ml-evs reviewed (approved):
These are not blocking my approval, and I will resolve them immediately after posting. Let's save fixes like this for a future PR now.

vaitkus commented on schemas/src/defs/v1.2/properties/optimade/structures/space_group_symbol_hall.yaml:
```diff
@@ -0,0 +1,46 @@
+$$schema: "https://schemas.optimade.org/meta/v1.2/optimade/property_definition"
+$id: "https://schemas.optimade.org/defs/v1.2/properties/optimade/structures/space_group_symbol_hall"
+title: "hall space group symbol"
```
```suggestion
title: "Hall space group symbol"
```

rartino replied:
This may go quicker if I didn't change things that were already right :)

I also fixed "Cartesian site positions", because that Cartesian should be capitalized, right? (At least wikipedia seems to suggest so).

ml-evs commented:
🎉
